src/key_io.cpp

```diff
@@ -83,14 +83,22 @@ class DestinationEncoder
 
 CTxDestination DecodeDestination(const std::string& str, const CChainParams& params, std::string& error_str, std::vector<int>* error_locations)
 {
-    std::vector<unsigned char> data;
-    uint160 hash;
     error_str = "";
 
-    // Note this will be false if it is a valid Bech32 address for a different network
-    bool is_bech32 = (ToLower(str.substr(0, params.Bech32HRP().size())) == params.Bech32HRP());
+    static const uint8_t MAX_BASE58_CHARS = 100;
+    static const uint8_t MAX_BASE58_CHECK_CHARS = 21;
+
+    std::vector<unsigned char> data, bech32_data;
+
+    auto [bech32_encoding, bech32_hrp, bech32_chars] = bech32::Decode(str);
+    auto [bech32_error, bech32_error_loc] = bech32::LocateErrors(str);
+    bool is_bech32 = bech32_encoding != bech32::Encoding::INVALID;
 
-    if (!is_bech32 && DecodeBase58Check(str, data, 21)) {
+    auto check_base58 = [&]() { return DecodeBase58Check(str, data, MAX_BASE58_CHECK_CHARS); };
+    
+
+    if (!is_bech32 && check_base58()) {
```

Comment:
as mentioned, if we invert the condition we'd have `if (is_bech32) { ... } else if (check_base58()).
But I'm not even sure why these two are extracted, I don't find the `check_base58` or `is_bech32` to be a helpful abstraction here - I'd inline them

Also, bech32 and base58 shouldn't be handled in the same method, they're not the same responsibility - could you please extract their handling to smaller, more focused methods?

I'd test and extract them in separate commits as well, I don't see a reason for combining them in any way.

Reply:
In 04d5d4acab4259505654a506c838fecb29fdb88d _Base58 decoding logic + bech32 decoding network awareness_: splitting bech32 and base58 handling in an earlier commit might also help to focus the current commit more on base58. I still a lot of lines touching bech32 code.